auth/standard_role_manager.cc

```diff
@@ -278,6 +287,11 @@ future<> standard_role_manager::stop() {
     return _stopped.handle_exception_type([] (const sleep_aborted&) { }).handle_exception_type([](const abort_requested_exception&) {});;
 }
 
+future<> standard_role_manager::ensure_superuser_is_created() {
+    co_await get_units(_superuser_created, 1);
```

Comment:
`semaphore` is the wrong tool for the job. What you want is a `promise<>` + `future<>` pair.

Reply:
> it has the correct semantics: it can be accessed multiple times

if `ensure_superuser_is_created()` is called more than once, we're in trouble: every call other than the first will **hang**.

With promise+future, the second call will crash on an assertion (IIRC). I think that is better than a hang. Definitely easier to debug.

Reply:
Ah, that's what it does... we get the units, and then immediately destroy them.

It's a bit hacky, don't you think?

`get_units` should probably be marked `[[nodiscard]]` and here you would have to explicitly discard them (by `(void)` or whatever), it would make it a bit clearer.

But still

IIUC `ensure_superuser_is_created()` should still be called exactly once -- when we initialize Scylla. Is it intended to be called more than once? If not, then wouldn't calling more than once be a bug (and bugs should be preferably catched as early as possible?)

Reply:
Either way, if we decide that "ensuring" should or can be done multiple times, the idiomatic way in seastar to have multiple waiters to wait for something to happen is `shared_promise`/`shared_future`, a semaphore where you get units only to immediately discard them is... weird.

Reply:
I think whatever is simplest wins, readability matters too. Semaphore is commonly used. It doesn't matter if functions calls are serialized - there is exactly one call in the code, and even if there will be 10, still doesn't matter.

By the way  condition_variable seems even more appropriate than suggested future+promise.

Reply:
> I think whatever is simplest wins, readability matters too

Agreed.

> Semaphore is commonly used

But not like that. Getting units from a semaphore just to destroy them immediately is far from common. It is abuse. Therefore it is not readable and not simple.